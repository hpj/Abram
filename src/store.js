const stores = {};

/**
* @param { string } name
* @param { {} } state
* @returns { Store }
*/
export function createStore(name, state)
{
  // eslint-disable-next-line security/detect-object-injection
  return (stores[name] = new Store(state));
}

/**
* @param { string } name
* @returns { Store }
*/
export function getStore(name)
{
  // eslint-disable-next-line security/detect-object-injection
  return stores[name];
}

/**
* @param { string } name
*/
export function deleteStore(name)
{
  // eslint-disable-next-line security/detect-object-injection
  delete stores[name];
}

export default class Store
{
  /**
  * @param { {} } state
  */
  constructor(state)
  {
    this.subscriptions = [];
    this.state = state || {};
  }

  /**
  * @param { import('react').Component } component
  * @returns { Store }
  */
  mount(component)
  {
    component.state = this.state;

    return this;
  }

  /**
  * @param { import('react').Component } component
  * @returns { Store }
  */
  subscribe(component)
  {
    if (component && component.setState && this.subscriptions.indexOf(component) < 0)
    {
      this.subscriptions.push(component);

      component.setState(this.state);

      return this;
    }

    return false;
  }

  /**
  * @param { import('react').Component } component
  * @returns { boolean }
  */
  unsubscribe(component)
  {
    const index = this.subscriptions.indexOf(component);

    if (index > -1)
    {
      this.subscriptions.splice(index, 1);

      return true;
    }

    return false;
  }

  /**
  * @param { {} } state
  * @returns { Store }
  */
  set(state)
  {
    this.state = {
      ...this.state,
      ...state
    };

    // dispatch changes
    this.dispatch();

    return this;
  }

  /**
  * @returns { Store }
  */
  dispatch()
  {
    this.subscriptions.forEach((component) =>
    {
      if (component && component.setState)
        component.setState(this.state);
    });

    return this;
  }
}