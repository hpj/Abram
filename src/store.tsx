/* eslint-disable security/detect-object-injection */
/* eslint-disable @typescript-eslint/ban-types */
/* eslint-disable @typescript-eslint/no-explicit-any */

import React from 'react';

let store: Store;

export function createStore(state: object): Store
{
  store = new Store(state);

  return store;
}

export function getStore(): Store
{
  return store;
}

export class StoreComponent<Props = {}, State = {}> extends React.Component<Props, State>
{
  constructor(state?: object)
  {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    //@ts-ignore
    super();
    
    // get store
    this.store = getStore();

    this.state = {
      ...state,
      ...this.store.state
    } as State;
  }

  store: Store;
  state: State;

  componentDidMount(): void
  {
    this.store.subscribe(this);
  }

  componentWillUnmount(): void
  {
    this.store.unsubscribe(this);
  }

  /** Emits before the state changes,
  * allows modification to the state before it's dispatched
  */
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  stateWillChange(newState: State): Partial<State> | void
  {
    //
  }

  /** Whitelist what changes are allowed to be dispatched to this component,
  * Improving performance (this is recommended, specially on large apps),
  * Not overriding this function will allow the component to receive any and all dispatches
  */
  // istanbul ignore next
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  stateWhitelist(changes: Partial<State>): boolean
  {
    return true;
  }

  /** Emits every time a new state gets dispatched
  */
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  stateDidChange(state: State, changes: Partial<State>, old: State): void
  {
    //
  }
}

export default class Store
{
  constructor(state?: object)
  {
    this.state = state ?? {};

    this.changes = {};
  }
  
  subscriptions: StoreComponent[] = []

  state: any

  changes: any

  mount(component: React.Component): Store
  {
    component.state = this.state;

    return this;
  }

  subscribe(component: React.Component): Store | undefined
  {
    if (component?.setState && this.subscriptions.indexOf(component as StoreComponent) < 0)
    {
      this.subscriptions.push(component as StoreComponent);

      component.setState(this.state);

      return this;
    }
  }

  unsubscribe(component: React.Component): boolean | undefined
  {
    const index = this.subscriptions.indexOf(component as StoreComponent);

    if (index > -1)
    {
      this.subscriptions.splice(index, 1);

      return true;
    }
  }

  set(state: object, callback?: () => void): Store
  {
    const keys = Object.keys(state);

    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    keys.forEach((key) => this.changes[key] = this.state[key] = state[key]);

    // this.changes = {
    //   ...this.changes,
    //   ...state
    // };

    // this.state = {
    //   ...this.state,
    //   ...state
    // };

    // dispatch changes
    this.dispatch().then(callback);

    return this;
  }

  async dispatch(): Promise<void>
  {
    /** transform all values of the new state to true
    */
    const booleanify = (obj: any): any =>
    {
      const out = {} as any;

      const keys = Object.keys(obj);

      keys.forEach(key => out[key] = true);

      // keys.forEach((key) =>
      // {
      //   if (!Array.isArray(obj[key]) && typeof obj[key] === 'object')
      //     out[key] = booleanify(obj[key]);
      //   else
      //     out[key] = true;
      // });

      return out;
    };
    
    const promises: Promise<void>[] = [];

    const changesFingerprint = booleanify(this.changes);

    this.subscriptions.forEach((component) =>
    {
      // callback to notify components when they state will change
      const modified = component.stateWillChange?.call(component, this.state);

      if (modified)
      {
        const keys = Object.keys(modified);

        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        keys.forEach(key => this.state[key] = modified[key]);

        // this.state = {
        //   ...this.state,
        //   ...modified
        // };
      }
    });

    this.subscriptions.forEach((component) =>
    {
      if (
        typeof component.stateWhitelist === 'function' &&
        component.stateWhitelist.call(component, changesFingerprint) !== true
      )
        return;
      
      const old = { ...component.state };

      promises.push(new Promise((resolve) =>
      {
        component.setState(this.state, () =>
        {
          component.stateDidChange?.call(component, this.state, this.changes, old);
          
          resolve();
        });
      }));
    });

    this.changes = {};

    await Promise.all(promises);
  }
}