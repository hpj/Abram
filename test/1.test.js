import React from 'react';

import { render, fireEvent } from 'react-native-testing-library';

import MockDate from 'mockdate';

import { createStore, deleteStore } from '../src/store.js';

import App from '../src/app.js';

beforeEach(() =>
{
  createStore('app', { index: 0 });

  // https://stackoverflow.com/a/51067606/10336604

  MockDate.set(0);

  jest.useFakeTimers();
});

afterEach(() =>
{
  deleteStore('app');
});

describe('<App/>', () =>
{
  test('renders correctly', () =>
  {
    const component = render(<App/>);

    expect(component.toJSON()).toMatchSnapshot();

    // simulate pressing the discover navigation button
    fireEvent.press(component.getByTestId('bn-discover'));

    // finish the view transition animation
    global.timeTravel(150);

    expect(component.toJSON()).toMatchSnapshot();
  });
});