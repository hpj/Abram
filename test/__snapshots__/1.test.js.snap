// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App/> renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#000000",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      pointerEvents="box-none"
      style={
        Object {
          "backgroundColor": "red",
          "bottom": 0,
          "left": 0,
          "opacity": 1,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "transform": Array [
            Object {
              "scaleX": 1,
            },
            Object {
              "scaleY": 1,
            },
          ],
        }
      }
    />
    <View
      pointerEvents="none"
      style={
        Object {
          "backgroundColor": "green",
          "bottom": 0,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "transform": Array [
            Object {
              "scaleX": 0.8,
            },
            Object {
              "scaleY": 0.8,
            },
          ],
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#000000",
        "flexDirection": "row",
        "height": 56,
        "justifyContent": "center",
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
            "maxWidth": 168,
            "minWidth": 80,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "flex-end",
              "backgroundColor": "#4B4B4B",
              "borderRadius": 44,
              "height": 44,
              "position": "absolute",
              "width": 44,
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "#ffffff",
                "borderRadius": 12,
                "height": 12,
                "width": 12,
              }
            }
          />
        </View>
        <Text />
        <View />
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
            "maxWidth": 168,
            "minWidth": 80,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "flex-end",
              "backgroundColor": "rgba(75,75,75,0.3)",
              "borderRadius": 44,
              "height": 44,
              "position": "absolute",
              "width": 44,
            }
          }
        >
          <View />
        </View>
        <Text />
        <View />
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;

exports[`<App/> renders correctly 2`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#000000",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      pointerEvents="none"
      style={
        Object {
          "backgroundColor": "red",
          "bottom": 0,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "transform": Array [
            Object {
              "scaleX": 0.8,
            },
            Object {
              "scaleY": 0.8,
            },
          ],
        }
      }
    />
    <View
      pointerEvents="box-none"
      style={
        Object {
          "backgroundColor": "green",
          "bottom": 0,
          "left": 0,
          "opacity": 1,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "transform": Array [
            Object {
              "scaleX": 1,
            },
            Object {
              "scaleY": 1,
            },
          ],
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#000000",
        "flexDirection": "row",
        "height": 56,
        "justifyContent": "center",
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
            "maxWidth": 168,
            "minWidth": 80,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "flex-end",
              "backgroundColor": "#4B4B4B",
              "borderRadius": 44,
              "height": 44,
              "position": "absolute",
              "width": 44,
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "#ffffff",
                "borderRadius": 12,
                "height": 12,
                "width": 12,
              }
            }
          />
        </View>
        <Text />
        <View />
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
            "maxWidth": 168,
            "minWidth": 80,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "flex-end",
              "backgroundColor": "rgba(75,75,75,0.3)",
              "borderRadius": 44,
              "height": 44,
              "position": "absolute",
              "width": 44,
            }
          }
        >
          <View />
        </View>
        <Text />
        <View />
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
